fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -193,18 +199,70 @@ public void afterHistory(FHIRPersistenceEvent event) throws FHIRPersistenceInter
+                    Resource resource = entry.getResource();
+                    if (resource instanceof Provenance) {
+                        // skip enforcement of provenance resources, but keep track of their targets
+                        provenanceTargets.add(((Provenance) resource).getTarget());
+                    } else {
+                        enforce(resource, getPatientIdFromToken(jwt), Permission.READ, getScopesFromToken(jwt));
+                        relativeResourcePaths.add(resource.getClass().getSimpleName() + "/" + resource.getId());
+                    }
                 }
             }
+
+            enforceProvenance(provenanceTargets, relativeResourcePaths);
         } else {
             throw new IllegalStateException("Expected event resource of type Bundle but found " +
                     event.getFhirResource().getClass().getSimpleName());
         }
     }
 
+    /**
+     * If the search result contains one or more Provenance resources,
+     * then reject the request unless each Provenance targets at least one resource in the response bundle
+     *
+     * @throws IllegalStateException if the baseUrl cannot be computed from the request context
+     * @throws FHIRPersistenceInterceptorException if the list of provenanceTargets contains
+     */
+    private void enforceProvenance(List<List<Reference>> provenanceTargets, Set<String> relativeResourcePaths)
+            throws FHIRPersistenceInterceptorException {
+        if (provenanceTargets.isEmpty()) return;
+
+        String baseUrl;
+        try {
+            baseUrl = ReferenceUtil.getServiceBaseUrl();
+        } catch (FHIRSearchException e) {
+            throw new IllegalStateException("Unexpected error while computing the service baseUrl for ");
+        }
+
+        for (List<Reference> targets : provenanceTargets) {
+            boolean allow = false;
+            for (Reference r : targets) {
+                ReferenceValue referenceValue = ReferenceUtil.createReferenceValueFrom(r, baseUrl);
+                if (ReferenceValue.ReferenceType.LITERAL_RELATIVE == referenceValue.getType()) {
+                    if (relativeResourcePaths.contains(referenceValue.getTargetResourceType() + "/" + referenceValue.getValue())) {
+                        allow = true;
+                        break;
+                    }
+                } else if (log.isLoggable(Level.FINE)){
+                    log.fine("Skipping target '" + referenceValue.getValue() + "' of type '" + referenceValue.getType() +
+                            "' during enforcement of Provenance access.");
+                }
+            }
+
+            if (!allow) {
+                throw new FHIRPersistenceInterceptorException(REQUEST_NOT_PERMITTED)
+                        .withIssue(FHIRUtil.buildOperationOutcomeIssue(REQUEST_NOT_PERMITTED, IssueType.FORBIDDEN));
+            }
+        }
+    }
+
```

Comment:
Question on one other scenario: what if a _revinclude search is done on a resource type that is not in the patient compartment, such that a resource and its related Provenance resource are returned. For example, a Practitioner and Provenance resource are returned. We'll allow the Practitioner resource, because we allow non-patient compartment resources. And thus we'll allow the Provenance resource, since it targets the Practitioner. But should it be allowed since it is technically not associated with the patient?

Reply:
I had the same question.  I was leaning toward only allowing Provenance resources which target a Patient-compartment resource, but that would require an extra line or two of code and I really don't know which way is "right".  As implemented, we basically leave it up to the IG to determine on which resource types, a `_revinclude:Provenance:target` is allowed

Reply:
Works for me.